src/main/java/apoc/hashing/Fingerprinting.java

```diff
@@ -0,0 +1,198 @@
+            if (o == null || getClass() != o.getClass()) return false;
+
+            EndNodeRelationshipHashTuple that = (EndNodeRelationshipHashTuple) o;
+
+            if (endNodeHash != null ? !endNodeHash.equals(that.endNodeHash) : that.endNodeHash != null) return false;
+            return relationshipHash != null ? relationshipHash.equals(that.relationshipHash) : that.relationshipHash == null;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = endNodeHash != null ? endNodeHash.hashCode() : 0;
+            result = 31 * result + (relationshipHash != null ? relationshipHash.hashCode() : 0);
+            return result;
+        }
+
+        public String getEndNodeHash() {
+            return endNodeHash;
+        }
+
+        public String getRelationshipHash() {
+            return relationshipHash;
+        }
+    }
+
+    private String fingerprintNode(Node node, List<String> excludedPropertyKeys) {
+        String s = withMessageDigest(md -> {
+            StreamSupport.stream(node.getLabels().spliterator(), false)
+                    .map(Label::name)
+                    .sorted()
+                    .map(String::getBytes)
+                    .forEach(md::update);
+            addPropertiesToDigest(node, excludedPropertyKeys, md);
+        });
+//        System.out.println("md " + s + " for node " + node.getLabels() + " " + node.getAllProperties());
+        return s;
+    }
+
+    private String fingerprintRelationship(Relationship rel, List<String> excludedPropertyKeys) {
+        String s = withMessageDigest(md -> {
+            md.update(rel.getType().name().getBytes());
+            addPropertiesToDigest(rel, excludedPropertyKeys, md);
+        });
+//        System.out.println("md " + s + " for rel " + rel.getType() + " " + rel.getAllProperties());
+        return s;
+
+    }
+
+    private void addPropertiesToDigest(PropertyContainer propertyContainer, List<String> excludedPropertyKeys, MessageDigest md) {
+        propertyContainer.getAllProperties().entrySet().stream()
+                .filter(e -> !excludedPropertyKeys.contains(e.getKey()))
+                .sorted(Map.Entry.comparingByKey())
+                .forEachOrdered(entry -> {
+                    md.update(entry.getKey().getBytes());
+                    md.update(convertValueToString(entry.getValue()).getBytes());
+                });
+    }
+
+    private String withMessageDigest(Consumer<MessageDigest> consumer) {
+        try {
+            MessageDigest md = MessageDigest.getInstance(DIGEST_ALGORITHM);
+            consumer.accept(md);
+            return DatatypeConverter.printHexBinary(md.digest());
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private String convertValueToString(Object value) {
+        if (value.getClass().isArray()) {
+            return nativeArrayToString(value);
+        } else {
+            return value.toString();
```

Comment:
what about nested values / iterables / maps?
And would it be more efficient to actually add the value's bytes to the MD?

Reply:
nesting should work now. I've refactored the code to work recursively.